pkg/image/controller/controller.go

```diff
@@ -29,8 +122,77 @@ func needsImport(stream *api.ImageStream) bool {
 	return stream.Annotations == nil || len(stream.Annotations[api.DockerImageRepositoryCheckAnnotation]) == 0
 }
 
-// retryCount is the number of times to retry on a conflict when updating an image stream
-const retryCount = 2
+func (c *ImportController) handleImport() {
+	for {
+		select {
+		case <-c.stopChan:
+			return
+
+		case staleImageStream := <-c.work:
+			glog.V(1).Infof("popped %s from the worklist", workingSetKey(staleImageStream))
+
+			c.importImageStream(staleImageStream)
+		}
+	}
+}
+
+func (c *ImportController) importImageStream(staleImageStream *api.ImageStream) {
+	// if we're already in the workingset, that means that some thread is already trying to do an import for this.
+	// This does NOT mean that we shouldn't attempt to do this work, only that we shouldn't attempt to do it now.
+	if !c.addToWorkingSet(staleImageStream) {
+		// If there isn't any other work in the queue, wait for a while so that we don't hot loop.
+		// Then requeue to the end of the channel.  That allows other work to continue without delay
+		if len(c.work) == 0 {
+			time.Sleep(100 * time.Millisecond)
+		}
+		glog.V(5).Infof("requeuing %s to the worklist", workingSetKey(staleImageStream))
+		c.work <- staleImageStream
+
+		return
+	}
+	defer c.removeFromWorkingSet(staleImageStream)
+
+	err := kclient.RetryOnConflict(kclient.DefaultBackoff, func() error {
+		liveImageStream, err := c.streams.ImageStreams(staleImageStream.Namespace).Get(staleImageStream.Name)
+		if err != nil {
+			return err
+		}
+		if !needsImport(liveImageStream) {
+			return nil
+		}
+
+		// if we're notified, do work and then start waiting again.
+		return c.Next(liveImageStream)
```

Comment:
> the expensive work is in Next... looking up the tags from the remote registry... I hate to redo that work because of a conflict on update

Are you talking about:
 1. The error case of the error case, in which case I'm hard pressed to care, but I'll suppress it if you like.  (double errors are a small edge)
 1. The success case, in which case I blindly updating the object is wrong now because there's no diff conflict checking and since `spec.dockerImageRepository` is mutable so you need a compatibility test, and we either support (or want to support) selectively choosing which tags to import which means you need a covers test as well.  Leading me to item 3
 1. I can definitely refactor this controller into a shape more to my (our?) liking, but its on @smarterclayton's list already.  I'd hate to go to the effort only to have it discarded in a couple weeks.  I'm hoping to keep this small and simply enable 10 imports to run at once.  I figured @smarterclayton would make it beautiful.